modules/tensor_mechanics/src/materials/crystal_plasticity/CrystalPlasticityStressUpdateBase.C

```diff
@@ -0,0 +1,400 @@
+    _slip_incr_tol(getParam<Real>("slip_increment_tolerance")),
+    _resistance_tol(getParam<Real>("resistance_tol")),
+    _zero_tol(getParam<Real>("zero_tol")),
+
+    _slip_resistance(declareProperty<std::vector<Real>>(_base_name + "slip_resistance")),
+    _slip_resistance_old(getMaterialPropertyOld<std::vector<Real>>(_base_name + "slip_resistance")),
+    _slip_increment(declareProperty<std::vector<Real>>(_base_name + "slip_increment")),
+
+    _slip_direction(_number_slip_systems * LIBMESH_DIM),
+    _slip_plane_normal(_number_slip_systems * LIBMESH_DIM),
+    _flow_direction(declareProperty<std::vector<RankTwoTensor>>(_base_name + "flow_direction")),
+    _tau(declareProperty<std::vector<Real>>(_base_name + "applied_shear_stress"))
+{
+  _substep_dt = 0.0;
+
+  getSlipSystems();
+  sortCrossSlipFamilies();
+
+  if (parameters.isParamSetByUser("number_cross_slip_directions"))
+    _calculate_cross_slip = true;
+  else
+    _calculate_cross_slip = false;
+}
+
+void
+CrystalPlasticityStressUpdateBase::initQpStatefulProperties()
+{
+  _tau[_qp].resize(_number_slip_systems);
+
+  _flow_direction[_qp].resize(_number_slip_systems);
+  for (unsigned int i = 0; i < _number_slip_systems; ++i)
+  {
+    _flow_direction[_qp][i].zero();
+    _tau[_qp][i] = 0.0;
+  }
+
+  _slip_resistance[_qp].resize(_number_slip_systems);
+  _slip_increment[_qp].resize(_number_slip_systems);
+}
+
+void
+CrystalPlasticityStressUpdateBase::getSlipSystems()
+{
+  bool orthonormal_error = false;
+
+  getPlaneNormalAndDirectionVectors(_slip_sys_file_name,
+                                    _number_slip_systems,
+                                    _slip_plane_normal,
+                                    _slip_direction,
+                                    orthonormal_error);
+
+  if (orthonormal_error)
+    mooseError("CrystalPlasticityStressUpdateBase Error: The slip system file contains a slip "
+               "direction and "
+               "plane normal pair that are not orthonormal");
+}
+
+void
+CrystalPlasticityStressUpdateBase::getPlaneNormalAndDirectionVectors(
+    const FileName & vector_file_name,
+    const unsigned int & number_dislocation_systems,
+    DenseVector<Real> & plane_normal_vector,
+    DenseVector<Real> & direction_vector,
+    bool & orthonormal_error)
+{
+  Real vec[LIBMESH_DIM];
+  std::ifstream fileslipsys;
+
+  MooseUtils::checkFileReadable(vector_file_name);
+
+  fileslipsys.open(vector_file_name.c_str());
+
+  for (unsigned int i = 0; i < number_dislocation_systems; ++i)
+  {
+    // Read the plane normal
+    for (unsigned int j = 0; j < LIBMESH_DIM; ++j)
+      if (!(fileslipsys >> vec[j]))
+        mooseError(
+            "CrystalPlasticityStressUpdateBase Error: Premature end of file reading plane normal "
+            "vectors from the file ",
+            vector_file_name);
+
+    // Normalize the vectors
+    Real magnitude;
+    magnitude = Utility::pow<2>(vec[0]) + Utility::pow<2>(vec[1]) + Utility::pow<2>(vec[2]);
```

Comment:
You size `vec` to `LIBMESH_DIM` but access `[2]` regardless. I'd add a static assert to make sure LIBMESH_DIM == 3 at the least.